consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java

```diff
@@ -0,0 +1,137 @@
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * This consistent {@link Sampler} adjust the sampling probability dynamically to limit the rate of
+ * sampled spans.
+ *
+ * <p>This sampler uses exponential smoothing to estimate on irregular data (compare Wright, David
+ * J. "Forecasting data published at irregular time intervals using an extension of Holt's method."
+ * Management science 32.4 (1986): 499-510.) to estimate the current rate of spans.
+ */
+@Immutable
+public class ConsistentRateLimitingSampler extends ConsistentSampler {
+
+  private final String description;
+  private final LongSupplier nanoTimeSupplier;
+  private final double inverseAdaptationTimeNanos;
+  private final double targetSpansPerNanosLimit;
+
+  @GuardedBy("this")
+  private double effectiveWindowCount;
+
+  @GuardedBy("this")
+  private double effectiveWindowNanos;
+
+  @GuardedBy("this")
+  private long lastNanoTime;
+
+  /**
+   * Constructor.
+   *
+   * @param targetSpansPerSecondLimit the desired spans per second limit
+   * @param adaptationTimeSeconds the typical time to adapt to a new load (time constant used for
+   *     exponential smoothing)
+   */
+  public ConsistentRateLimitingSampler(
+      double targetSpansPerSecondLimit, double adaptationTimeSeconds) {
+    this(
+        targetSpansPerSecondLimit,
+        adaptationTimeSeconds,
+        DefaultRandomGenerator.get(),
+        System::nanoTime);
+  }
+
+  /**
+   * Constructor.
+   *
+   * @param targetSpansPerSecondLimit the desired spans per second limit
+   * @param adaptationTimeSeconds the typical time to adapt to a new load (time constant used for
+   *     exponential smoothing)
+   * @param threadSafeRandomGenerator a thread-safe random generator
+   * @param nanoTimeSupplier a supplier for the current nano time
+   */
+  public ConsistentRateLimitingSampler(
+      double targetSpansPerSecondLimit,
+      double adaptationTimeSeconds,
+      RandomGenerator threadSafeRandomGenerator,
+      LongSupplier nanoTimeSupplier) {
+    super(threadSafeRandomGenerator);
+
+    if (targetSpansPerSecondLimit < 0.0) {
+      throw new IllegalArgumentException("Limit for sampled spans per second must be nonnegative!");
+    }
+    if (adaptationTimeSeconds < 0.0) {
+      throw new IllegalArgumentException("Adaptation rate must be nonnegative!");
+    }
+    this.description =
+        String.format(
+            "ConsistentRateLimitingSampler{%.6f, %.6f}",
+            targetSpansPerSecondLimit, adaptationTimeSeconds);
+    this.nanoTimeSupplier = requireNonNull(nanoTimeSupplier);
+
+    this.inverseAdaptationTimeNanos = 1e-9 / adaptationTimeSeconds;
+    this.targetSpansPerNanosLimit = 1e-9 * targetSpansPerSecondLimit;
+
+    synchronized (this) {
+      this.effectiveWindowCount = 0;
+      this.effectiveWindowNanos = 0;
+      this.lastNanoTime = nanoTimeSupplier.getAsLong();
+    }
+  }
+
+  private synchronized double updateAndGetSamplingProbability() {
```

Comment:
Any chances this algorithm can be adapted to support "worst-case" guarantee, such as a predefined confidence level that the specified rate will not be exceeded?

Reply:
To calculate such a "confidence level", a model is needed that describes when the spans occur. The simplest model is probably the Poisson process, which assumes that the rate is constant and that all spans are independent. However, both assumptions are rarely true in practice. The rate is usually not constant, and the occurrence of spans is also not independent (consider, for example, a trace that always calls a service twice, so spans are always generated in pairs). 

We also need to specify how the rate limit is defined. Usually this is done in terms of a time interval. E.g., a maximum of 200 spans in each 1-minute interval.  

When everything is well defined, the model, the rate limit definition, and the adaptive sampling algorithm, the expected frequency of rate limit violations can be calculated or simulated. I assume this is what you are ultimately interested in?

Reply:
Yes, I believe an algorithm that gives some kind of confidence level is a good compromise between a purely probabilistic approach and any algorithm based on "leaky bucket" approach. I believe it can be made almost-unbiased and almost-rate-limiting.
BTW, your algorithm, when faced with spans coming in pairs (at almost the same point of time) can reduce the sampling probability for the second span significantly.

Reply:
> BTW, your algorithm, when faced with spans coming in pairs (at almost the same point of time) can reduce the sampling probability for the second span significantly.

Yes, this is true if the parameter for the adaptation time parameter is chosen small (in the order of magnitude of the time span between subsequent pairs or less).

This algorithm was motivated by a scenario where the span rate is suddenly increased. In this case, the algorithm immediately starts to decrease the sampling probability. In contrast, an adaptive sampler that adjusts the sampling rate periodically (e.g., based on the estimated rate of the previous period) would take much longer and the desired sampling rate would be exceeded much further.